serverless/cmd/integrate/main.go

```diff
@@ -18,48 +18,125 @@ package main
+	privKeyFile = flag.String("private_key", "", "Location of private key file. If unset, uses the contents of the SERVERLESS_LOG_PRIVATE_KEY environment variable.")
 )
 
 func main() {
 	flag.Parse()
 	h := hasher.DefaultHasher
+	// Read log public key from file or environment variable
+	var pubKey string
+	if len(*pubKeyFile) > 0 {
+		k, err := ioutil.ReadFile(*pubKeyFile)
+		if err != nil {
+			glog.Exitf("failed to read public_key file: %q", err)
+		}
+		pubKey = string(k)
+	} else {
+		pubKey = os.Getenv("SERVERLESS_LOG_PUBLIC_KEY")
+		if len(pubKey) == 0 {
+			glog.Exit("supply public key file path using --public_key or set SERVERLESS_LOG_PUBLIC_KEY environment variable")
+		}
+	}
+	// Read log private key from file or environment variable
+	var privKey string
+	if len(*privKeyFile) > 0 {
+		k, err := ioutil.ReadFile(*privKeyFile)
+		if err != nil {
+			glog.Exitf("failed to read private_key file: %q", err)
+		}
+		privKey = string(k)
+	} else {
+		privKey = os.Getenv("SERVERLESS_LOG_PRIVATE_KEY")
+		if len(privKey) == 0 {
+			glog.Exit("supply private key file path using --private_key or set SERVERLESS_LOG_PUBLIC_KEY environment variable")
+		}
+	}
+
+	var cpNote note.Note
+	s, err := note.NewSigner(privKey)
+	if err != nil {
+		glog.Exitf("failed to instantiate signer: %q", err)
+	}
+
+	if *initialise {
+		st, err := fs.Create(*storageDir, h.EmptyRoot())
+		if err != nil {
+			glog.Exitf("failed to create log: %q", err)
+		}
+		cp := st.Checkpoint()
+		err = signAndWrite(&cp, cpNote, s, st)
+		if err != nil {
+			glog.Exitf("failed to sign: %q", err)
+		}
+		os.Exit(0)
+	}
 
 	// init storage
 	cpRaw, err := fs.ReadCheckpoint(*storageDir)
 	if err != nil {
-		glog.Exitf("Failed to read log checkpoint: %q", err)
+		glog.Exitf("failed to read log checkpoint: %q", err)
+	}
+
+	// Check signatures
+	v, err := note.NewVerifier(pubKey)
+	if err != nil {
+		glog.Exitf("Failed to instantiate Verifier: %q", err)
+	}
+	vCp, err := note.Open(cpRaw, note.VerifierList(v))
+	if err != nil {
+		glog.Exitf("failed to open Checkpoint: %q", err)
 	}
 	var cp fmtlog.Checkpoint
-	if _, err := cp.Unmarshal(cpRaw); err != nil {
-		glog.Exitf("Failed to unmarshal checkpoint: %q", err)
+	if _, err := cp.Unmarshal([]byte(vCp.Text)); err != nil {
+		glog.Exitf("failed to unmarshal checkpoint: %q", err)
 	}
 	st, err := fs.Load(*storageDir, &cp)
 	if err != nil {
-		glog.Exitf("Failed to load storage: %q", err)
+		glog.Exitf("failed to load storage: %q", err)
 	}
 
 	// Integrate new entries
 	newCp, err := log.Integrate(st, h)
 	if err != nil {
-		glog.Exitf("Failed to integrate: %q", err)
+		glog.Exitf("failed to integrate: %q", err)
 	}
 	if newCp == nil {
-		glog.Exit("Nothing to integrate")
+		glog.Exit("nothing to integrate")
 	}
 
-	// Persist new log checkpoint.
-	if err := st.WriteCheckpoint(newCp.Marshal()); err != nil {
-		glog.Exitf("Failed to store new log checkpoint: %q", err)
+	err = signAndWrite(newCp, cpNote, s, st)
```

Comment:
same inline `if err := ...; err != nil { ...` here too

Reply:
donne

Reply:
not done?